File: email-aggregator-go/src/syncsvc/statemachine.go
// Package syncsvc 实现同步编排器与 7 态同步状态机（深化文档 §13）。
// 状态机为纯函数式转换，非法迁移显式报错，可测试、可审计。
package syncsvc

import "errors"

// SyncState 同步状态
type SyncState string

const (
	StateUnconnected SyncState = "UNCONNECTED"
	StateAuthorizing SyncState = "AUTHORIZING"
	StateInitialFull SyncState = "INITIAL_FULL"
	StateIncremental SyncState = "INCREMENTAL"
	StateThrottled   SyncState = "THROTTLED"
	StateError       SyncState = "ERROR"
	StatePaused      SyncState = "PAUSED"
)

// SyncEvent 驱动状态迁移的事件
type SyncEvent string

const (
	EvAuthorize SyncEvent = "AUTHORIZE"
	EvFullDone  SyncEvent = "FULL_DONE"
	EvTick      SyncEvent = "TICK"
	EvThrottle  SyncEvent = "THROTTLE"
	EvResume    SyncEvent = "RESUME"
	EvFail      SyncEvent = "FAIL"
	EvPause     SyncEvent = "PAUSE"
	EvReset     SyncEvent = "RESET"
)

// transitions 显式状态迁移表（白名单；不在表内即非法）
var transitions = map[SyncState]map[SyncEvent]SyncState{
	StateUnconnected: {EvAuthorize: StateAuthorizing},
	StateAuthorizing: {EvFullDone: StateInitialFull, EvFail: StateError},
	StateInitialFull: {EvFullDone: StateIncremental, EvThrottle: StateThrottled, EvFail: StateError, EvPause: StatePaused},
	StateIncremental: {EvThrottle: StateThrottled, EvFail: StateError, EvPause: StatePaused, EvReset: StateInitialFull},
	StateThrottled:   {EvResume: StateIncremental, EvFail: StateError, EvPause: StatePaused},
	StateError:       {EvReset: StateInitialFull, EvResume: StateAuthorizing, EvPause: StatePaused},
	StatePaused:      {EvResume: StateIncremental, EvReset: StateInitialFull},
}

// CanTransition 查询 from 经 ev 能否到达 next
func CanTransition(from SyncState, ev SyncEvent) (SyncState, bool) {
	next, ok := transitions[from][ev]
	return next, ok
}

// Advance 推进状态机，非法迁移返回错误（不静默）
func Advance(from SyncState, ev SyncEvent) (SyncState, error) {
	if next, ok := CanTransition(from, ev); ok {
		return next, nil
	}
	return from, errors.New("illegal transition: " + string(from) + " --" + string(ev) + "-->")
}

File: email-aggregator/src/sync/stateMachine.ts
/**
 * 同步编排状态机（深化文档 §A.2，7 态）。
 * 由 sync-orchestrator 驱动；错误进入退避，成功回到守护态。
 * 采用显式迁移表，非法迁移抛错 —— 让"状态边界"可被测试与审计。
 */

export type SyncState =
  | "UNCONNECTED"
  | "AUTHORIZING"
  | "INITIAL_FULL"
  | "INCREMENTAL"
  | "THROTTLED"
  | "ERROR"
  | "PAUSED";

export type SyncEvent =
  | "AUTH_OK"
  | "FULL_DONE"
  | "CHANGE_RECEIVED"
  | "THROTTLED"
  | "TRANSIENT_ERR"
  | "FATAL_ERR"
  | "RESUME"
  | "PAUSE";

const TRANSITIONS: Record<SyncState, Partial<Record<SyncEvent, SyncState>>> = {
  UNCONNECTED: { AUTH_OK: "AUTHORIZING" },
  AUTHORIZING: { AUTH_OK: "INITIAL_FULL" },
  INITIAL_FULL: { FULL_DONE: "INCREMENTAL", TRANSIENT_ERR: "THROTTLED", FATAL_ERR: "ERROR" },
  INCREMENTAL: {
    CHANGE_RECEIVED: "INCREMENTAL",
    THROTTLED: "THROTTLED",
    TRANSIENT_ERR: "THROTTLED",
    FATAL_ERR: "ERROR",
    PAUSE: "PAUSED",
  },
  THROTTLED: { RESUME: "INCREMENTAL", TRANSIENT_ERR: "ERROR" },
  ERROR: { RESUME: "INCREMENTAL" },
  PAUSED: { RESUME: "INCREMENTAL" },
};

export class SyncStateMachine {
  private state: SyncState = "UNCONNECTED";
  private listeners: Array<(from: SyncState, to: SyncState, ev: SyncEvent) => void> = [];

  get current(): SyncState {
    return this.state;
  }

  onTransition(fn: (from: SyncState, to: SyncState, ev: SyncEvent) => void): void {
    this.listeners.push(fn);
  }

  can(event: SyncEvent): boolean {
    return !!TRANSITIONS[this.state][event];
  }

  /** 触发迁移；返回是否发生状态变化 */
  send(event: SyncEvent): boolean {
    const next = TRANSITIONS[this.state][event];
    if (!next) {
      throw new Error(`illegal transition: ${this.state} --${event}--> ?`);
    }
    if (next === this.state) {
      // 自环（如 INCREMENTAL 收到 CHANGE_RECEIVED），仍通知监听者
      this.listeners.forEach((l) => l(this.state, this.state, event));
      return false;
    }
    const from = this.state;
    this.state = next;
    this.listeners.forEach((l) => l(from, next, event));
    return true;
  }

  reset(): void {
    this.state = "UNCONNECTED";
  }
}
